### PR TITLE
DB-12144 Add new index to sys.sysconglomerates on conglomeratenumber 

### PR DESCRIPTION
**Pull request by arnaud-splice**

### CODE REVIEW

**Comment by arnaud-splice** [comment]
jenkins please test branch @cdh6.3.0,skipTestsLongerThan2Minutes

**Comment by cloudspliceci** [comment]
TEST FAILED
http://nexus.splicemachine.com:8080/job/spliceengine-PR/job/Splice-PR-Branch/3111/

File: splice_machine/src/main/java/com/splicemachine/derby/impl/sql/catalog/upgrade/SpliceCatalogUpgradeScripts.java [hunk 1]
```diff
@@ -78,6 +78,7 @@ public SpliceCatalogUpgradeScripts(SpliceDataDictionary sdd, TransactionControll
         this.tc=tc;
 
         scripts = new ArrayList<>();
+        addUpgradeScript(baseVersion4, 2020, new UpgradeAddConglomerateNumberIndex(sdd, tc));
```
Comment (dgomezferro) [inline splice_machine/src/main/java/com/splicemachine/derby/impl/sql/catalog/upgrade/SpliceCatalogUpgradeScripts.java:81]: Are you sure this should be executed before the `UpgradeConglomerateTable` script? Have you tested a pre-1996 upgrade?

**Review by dgomezferro — COMMENTED** [review]
Change looks good to me, very clean, I only have a couple of questions.

There's a bunch of other DataDictionaryImpl methods that also use the `getDescriptorViaHeap` path and could become slow. Most have to do with dropping all schema objects. I opened https://splicemachine.atlassian.net/browse/DB-12224 to track it.

**Review by ascend1 — COMMENTED** [review]
I wonder if some metadata queries could be tuned to utilize this new index and gain some performance improvement. We probably should walk through them quickly.

**Review by martinrupp — APPROVED** [review]

**Review by ascend1 — APPROVED** [review]

**Review by carolp-503 — APPROVED** [review]
+1 once all comments have been answered to appropriate depth.

File: splice_machine/src/main/java/com/splicemachine/derby/impl/sql/catalog/upgrade/SpliceCatalogUpgradeScripts.java (hunk 1 above)
Comment (arnaud-splice) [reply splice_machine/src/main/java/com/splicemachine/derby/impl/sql/catalog/upgrade/SpliceCatalogUpgradeScripts.java:83]: Tested with 1990 and the upgrade went smoothly.

File: db-engine/src/main/java/com/splicemachine/db/impl/sql/catalog/DataDictionaryImpl.java [hunk 2]
```diff
@@ -5677,29 +5680,24 @@ public ConglomerateDescriptor getConglomerateDescriptor(long conglomerateNumber)
      */
     @Override
     public ConglomerateDescriptor[] getConglomerateDescriptors(long conglomerateNumber) throws StandardException{
+        // XXX Make this an index scan
```
[the remark this answers is not in the document]
Comment (arnaud-splice) [reply db-engine/src/main/java/com/splicemachine/db/impl/sql/catalog/DataDictionaryImpl.java]: done

**Comment by arnaud-splice** [comment]
jenkins please test branch @cdh6.3.0,skipTestsLongerThan2Minutes

File: db-engine/src/main/java/com/splicemachine/db/impl/sql/catalog/DataDictionaryImpl.java [hunk 3]
```diff
@@ -340,6 +340,10 @@ public void boot(boolean create,Properties startParams) throws StandardException
                     SYSCONGLOMERATESRowFactory.SYSCONGLOMERATES_INDEX3_ID,
                     getBootParameter(startParams,CFG_SYSCONGLOMERATES_INDEX3_ID,true));
 
+            coreInfo[SYSCONGLOMERATES_CORE_NUM].setIndexConglomerate(
+                    SYSCONGLOMERATESRowFactory.SYSCONGLOMERATES_INDEX4_ID,
+                    getBootParameter(startParams, CFG_SYSCONGLOMERATES_INDEX4_ID, false));
```
[the remark this answers is not in the document]
Comment (arnaud-splice) [reply db-engine/src/main/java/com/splicemachine/db/impl/sql/catalog/DataDictionaryImpl.java:353]: I've tried but hit a couple of issues. I created a separate item to address that: DB-12231

**Comment by cloudspliceci** [comment]
TEST FAILED
http://nexus.splicemachine.com:8080/job/spliceengine-PR/job/Splice-PR-Branch/3141/

**Comment by arnaud-splice** [comment]
jenkins please test branch @cdh6.3.0,skipTestsLongerThan2Minutes

**Comment by cloudspliceci** [comment]
TEST FAILED
http://nexus.splicemachine.com:8080/job/spliceengine-PR/job/Splice-PR-Branch/3147/

**Comment by martinrupp** [comment]
jenkins please test branch @cdh6.3.0,skipTestsLongerThan2Minutes

**Comment by cloudspliceci** [comment]
TEST FAILED
http://nexus.splicemachine.com:8080/job/spliceengine-PR/job/Splice-PR-Branch/3153/

**Comment by martinrupp** [comment]
jenkins please test branch @cdh6.3.0,skipTestsLongerThan2Minutes

**Comment by cloudspliceci** [comment]
TEST FAILED
http://nexus.splicemachine.com:8080/job/spliceengine-PR/job/Splice-PR-Branch/3155/

**Comment by arnaud-splice** [comment]
jenkins please test branch @cdh6.3.0,skipTestsLongerThan2Minutes

**Comment by cloudspliceci** [comment]
TEST FAILED
http://nexus.splicemachine.com:8080/job/spliceengine-PR/job/Splice-PR-Branch/3161/

**Comment by bklo94** [comment]
jenkins please test branch @cdh6.3.0

**Comment by cloudspliceci** [comment]
TEST SUCCEEDED +1

**Pull request merged**